<?xml version="1.0" encoding="utf-8"?>
<robot name="ranger_mini_v3">
    <link name="base_link">
    </link>

    <link name="fr_steering_wheel_link">
    </link>

    <joint name="fr_steering_joint" type="revolute">
        <origin rpy="0 0 0" xyz="0.25 -0.19 -0.1"/>
        <parent link="base_link"/>
        <child link="fr_steering_wheel_link"/>
        <axis rpy="0 0 0" xyz="0 0 -1"/>
        <limit effort="5" lower="-2.1" upper="2.1" velocity="6.28"/>
        <dynamics damping="5.0" friction="0.5"/>
    </joint>

    <link name="fr_wheel_link">
    </link>

    <joint name="fr_wheel" type="continuous">
        <origin rpy="0 0 0" xyz="0.0 0.0 -0.12"/>
        <parent link="fr_steering_wheel_link"/>
        <child link="fr_wheel_link"/>
        <axis rpy="0 0 0" xyz="0 1 0"/>
    </joint>

    <link name="rr_steering_wheel_link">
    </link>

    <joint name="rr_steering_joint" type="revolute">
        <origin rpy="0 0 0" xyz="-0.25 -0.19 -0.1"/>
        <parent link="base_link"/>
        <child link="rr_steering_wheel_link"/>
        <axis rpy="0 0 0" xyz="0 0 -1"/>
        <limit effort="5" lower="-2.1" upper="2.1" velocity="6.28"/>
        <dynamics damping="5.0" friction="0.5"/>
    </joint>

    <link name="rr_wheel_link">
    </link>

    <joint name="rr_wheel" type="continuous">
        <origin rpy="0 0 0" xyz="0.0 0.0 -0.12 "/>
        <parent link="rr_steering_wheel_link"/>
        <child link="rr_wheel_link"/>
        <axis rpy="0 0 0" xyz="0 1 0"/>
    </joint>

    <link name="rl_steering_wheel_link">
    </link>

    <joint name="rl_steering_joint" type="revolute">
        <origin rpy="0 0 0" xyz="-0.25 0.19 -0.1"/>
        <parent link="base_link"/>
        <child link="rl_steering_wheel_link"/>
        <axis rpy="0 0 0" xyz="0 0 -1"/>
        <limit effort="5" lower="-2.1" upper="2.1" velocity="6.28"/>
        <dynamics damping="5.0" friction="0.5"/>
    </joint>

    <link name="rl_wheel_link">
    </link>

    <joint name="rl_wheel" type="continuous">
        <origin rpy="0 0 0" xyz="0.0 0.0 -0.12 "/>
        <parent link="rl_steering_wheel_link"/>
        <child link="rl_wheel_link"/>
        <axis rpy="0 0 0" xyz="0 1 0"/>
    </joint>

    <link name="fl_steering_wheel_link">
    </link>

    <joint name="fl_steering_joint" type="revolute">
        <origin rpy="0 0 0" xyz="0.25 0.19 -0.1"/>
        <parent link="base_link"/>
        <child link="fl_steering_wheel_link"/>
        <axis rpy="0 0 0" xyz="0 0 -1"/>
        <limit effort="5" lower="-2.1" upper="2.1" velocity="6.28"/>
        <dynamics damping="5.0" friction="0.5"/>
    </joint>

    <link name="fl_wheel_link">
    </link>

    <joint name="fl_wheel" type="continuous">
        <origin rpy="0 0 0" xyz="0.0 0.0 -0.12"/>
        <parent link="fl_steering_wheel_link"/>
        <child link="fl_wheel_link"/>
        <axis rpy="0 0 0" xyz="0 1 0"/>
    </joint>


    <!-- xarm6 -->
    <link name="arm_base"/>

    <joint name="arm_to_ranger_joint" type="fixed">
        <parent link="arm_base"/>
        <child link="link_base"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
    </joint>

    <material name="Black">
        <color rgba="0.0 0.0 0.0 1.0"/>
    </material>
    <material name="Red">
        <color rgba="0.8 0.0 0.0 1.0"/>
    </material>
    <material name="White">
        <color rgba="1.0 1.0 1.0 1.0"/>
    </material>
    <material name="Silver">
        <color rgba="0.753 0.753 0.753 1.0"/>
    </material>

    <link name="link_base">
    </link>

    <link name="link1">
    </link>

    <joint name="joint1" type="revolute">
        <parent link="link_base"/>
        <child link="link1"/>
        <origin rpy="0 0 0" xyz="0 0 0.267"/>
        <axis xyz="0 0 1"/>
        <limit effort="50.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="3.14"/>
        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <link name="link2">
    </link>

    <joint name="joint2" type="revolute">
        <parent link="link1"/>
        <child link="link2"/>
        <origin rpy="-1.5708 0 0" xyz="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="50.0" lower="-2.059" upper="2.0944" velocity="3.14"/>
        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <link name="link3">
    </link>

    <joint name="joint3" type="revolute">
        <parent link="link2"/>
        <child link="link3"/>
        <origin rpy="0 0 0" xyz="0.0535 -0.2845 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="32.0" lower="-3.927" upper="0.19198" velocity="3.14"/>
        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <link name="link4">
    </link>

    <joint name="joint4" type="revolute">
        <parent link="link3"/>
        <child link="link4"/>
        <origin rpy="-1.5708 0 0" xyz="0.0775 0.3425 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="32.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="3.14"/>
        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <link name="link5">
    </link>

    <joint name="joint5" type="revolute">
        <parent link="link4"/>
        <child link="link5"/>
        <origin rpy="1.5708 0 0" xyz="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="32.0" lower="-1.69297" upper="3.141592653589793" velocity="3.14"/>
        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <link name="link6">
    </link>

    <joint name="joint6" type="revolute">
        <parent link="link5"/>
        <child link="link6"/>
        <origin rpy="-1.5708 0 0" xyz="0.076 0.097 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="20.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="3.14"/>
        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <link name="link_eef"/>

    <joint name="joint_eef" type="fixed">
        <origin rpy="0 0 0" xyz="0 0 0.172"/>
        <parent link="link6"/>
        <child link="link_eef"/>
    </joint>

    <!-- attach to ranger -->
    <joint name="ranger_with_xarm" type="fixed">
        <origin rpy="0 0 0" xyz="0.1855 0 0.028 "/>
        <parent link="base_link"/>
        <child link="arm_base"/>
    </joint>

</robot>